File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/icons/IconProvider.java
package ch.rmy.android.statusbar_tacho.icons;

import android.content.Context;
import android.content.res.Resources;

public class IconProvider {

    private static final int MAX_VALUE = 199;
    private static final String RES_FORMAT = "%03dicon";
    private static final String RES_TYPE = "drawable";

    private final Resources resources;
    private final String packageName;

    public IconProvider(Context context) {
        this.resources = context.getResources();
        this.packageName = context.getPackageName();
    }

    public int getIconForNumber(int number) {
        number = Math.max(0, Math.min(MAX_VALUE, number));
        String iconName = String.format(RES_FORMAT, number);
        return resources.getIdentifier(iconName, RES_TYPE, packageName);
    }

}

File: StatusBarSpeedometer/app/src/main/kotlin/ch/rmy/android/statusbar_tacho/icons/IconProvider.kt
package ch.rmy.android.statusbar_tacho.icons

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import androidx.annotation.DrawableRes
import java.util.Locale

class IconProvider(context: Context) {

    private val resources: Resources = context.resources
    private val packageName: String = context.packageName

    @SuppressLint("DiscouragedApi")
    @DrawableRes
    fun getIconForNumber(number: Int): Int =
        String.format(Locale.US, RES_FORMAT, number.coerceIn(0, MAX_VALUE))
            .let { iconName ->
                resources.getIdentifier(iconName, RES_TYPE, packageName)
            }

    companion object {

        private const val MAX_VALUE = 999
        private const val RES_FORMAT = "icon%03d"
        private const val RES_TYPE = "drawable"

    }

}
